lua/packer.lua

```diff
@@ -167,6 +168,108 @@ packer.use_rocks = function(rock)
   end
 end
 
+-- Merges new into old; returns merged spec.
+local function merge_specs(old, new)
+  local name = new.short_name
+
+  -- disable: Warn & ignore on disabling a previously enabled plugin
+  if old.disable then
+    if not new.disable then log.warn('Trying to disable previously non-disabled mod "' .. name .. '", ignoring') end
+  else
+    old.disable = new.disable
+  end
+
+  -- branch, tag, commit, lock: Must all be the same, else warning
+  for _, key in ipairs{'branch', 'commit', 'lock', 'tag'} do
+    if new[key] ~= old[key] then
+      log.warn('Plugin "' .. name .. '" wants to overwrite "' .. key .. '" setting. Keeping original value')
+    end
+  end
+
+  -- opt
+  -- - Conditions and opt_default have already been handled in manage()
+  -- - Rule of thumb: true < nil < false. Set the lowest necessary--if nil suffices, don't set false
+  -- - For manual_opt:
+  --   - If result is start, remove mo. Else, if omo or nmo, keep mo.
+  --   - There is no situation where opt = falsy but manual_opt = true
+  --
+  -- old new result (1 = true, - = nil, 0 = false, s = start, o = opt)
+  -- ---+---+------
+  --   -   -     -s
+  --   1   -     -s
+  --   -   1     -s
```

Comment:
I'm somewhat confused as to why these lines have the same result. For example, couldn't we have a case where `old.opt = true` because the "original" spec sets up some lazy-loading, but `new.opt = nil` because the new spec just customizes e.g. `config` and wants to leave the original lazy-loading intact?

Reply:
That would definitely make sense, yeah.

And that reveals one of the major pain points I've had while deciding on the policies for each setting: Sometimes more than one behavior may make sense, depending on the use-case, but I had to choose one for this PR. That only became apparent to me when I moved `merge_specs()` over to my own config, where deciding on those policies was surprisingly easy since I was the final arbiter on how these settings should behave.

This was actually another reason for me wanting to close this PR (one that I should have expressed explicitly, I apologize): In its current form, once you activate the `merge_duplicates` key, it might result in behavior that doesn't fit your use-case, and you cannot override it without another PR. This wouldn't be a problem if one could override how a key is handled, but that is blocked on migrating all the setting keys to handlers.

I feel this is a pretty big deal, so maybe we should in fact wait with merging this PR until the migration of the "official" keys to handlers has been made?

Reply:
That's a fair point, and if this PR no longer serves your needs, I understand if you don't want to dedicate more time to it - I should be able to take it over from here, if you'd prefer.

That said, one workaround is the `override` key I mentioned in another comment. Or, we can wait until I finally finish the refactoring to make everything handlers - maybe this will serve as good motivation for me to move that PR along....